src/dma/mod.rs

```diff
@@ -0,0 +1,726 @@
+                // This write can not fail since the DMA is not accessing the corresponding buffer
+                // yet or has hit the poison address and errored disabling itself.
+                unsafe {
+                    self.stream.set_memory_address(inactive, buf_ptr as usize);
+                }
+
+                // Acknowledge the TCIF.
+                self.stream.clear_transfer_complete_interrupt();
+
+                if single_buffer {
+                    unsafe {
+                        self.stream.enable();
+                    }
+                }
+
+                Ok(result)
+            }
+
+            /// Changes the buffer and restarts or continues a double buffer
+            /// transfer. This must be called immediately after a transfer complete
+            /// event. Returns the old buffer together with its `CurrentBuffer`. If an
+            /// error occurs, this method will return the old or new buffer with the error.
+            ///
+            /// This method will clear the transfer complete flag. Moreover, if
+            /// an overrun occurs, the stream will be disabled and the transfer error
+            /// flag will be set. This method can be called before the end of an ongoing
+            /// transfer only if not using double buffering, in that case, the current
+            /// transfer will be canceled and a new one will be started. A `NotReady`
+            /// error together with the new buffer will be returned if this method is called
+            /// before the end of a transfer while double buffering. A `SmallBuffer` error
+            /// together with the old buffer will be returned if the new buffer size does not
+            /// match the ongoing transfer size.
+            pub fn next_transfer(
+                &mut self,
+                new_buf: BUF,
+            ) -> Result<(BUF, CurrentBuffer, usize), DMAError> {
+                let mut buf = new_buf;
+                let mut last_remaining = 0usize;
+
+                let current =
+                    self.next_transfer_with(|mut old, current, remaining| {
+                        core::mem::swap(&mut old, &mut buf);
+                        last_remaining = remaining;
+                        (old, current)
+                    })?;
+                // TODO: return buf on Err
+                Ok((buf, current, last_remaining))
+            }
+
+            /// Clear half transfer interrupt (htif) for the DMA stream.
+            #[inline(always)]
+            pub fn clear_half_transfer_interrupt(&mut self) {
+                self.stream.clear_half_transfer_interrupt();
+            }
+
+            #[inline(always)]
+            pub fn get_half_transfer_flag(&self) -> bool {
+                STREAM::get_half_transfer_flag()
+            }
+        }
+    };
+}
+
+db_transfer_def!(DBTransfer, init, StaticWriteBuffer, write_buffer, mut;);
+db_transfer_def!(ConstDBTransfer, init_const, StaticReadBuffer, read_buffer;
```

Comment:
This actually precludes single transfers from using the `next_transfer()` API, which is still highly desirable even for single-buffer-mode, since you may want to continue using the same transfer with a new source/dest buffer (such is the case where asynchronous data collection from an external course)

Reply:
Note that we can currently work around this by falsely claiming those types of transfers are `DBTransfer` or `ConstDBTransfer`

Reply:
Is precluding single transfers from using `next_transfer()` highly desirable? Or not?

I think this my be a naming issue, because here "DB Transfer" means any transfer (including single buffered ones) happening on a DMA that supports double buffering. So that would be any transfer on DMA1, DMA2 or a BDMA. I see how this naming could be confusing, and would be solved by having a better collective description for transfers on (DMA1 or DMA2 or BDMAs).

Reply:
@ryan-summers is this comment still open? Other than this comment, I think we can merge the dma branch

Reply:
As pointed out, this actually works just fine. I was just confused by naming conventions.